### PR TITLE
rocAL Tensor Video pipeline changes

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Change instances of image to tensor
- Add changes to VideoFileSource, VideoFileResize and SequenceReader API
- Remove video loader module and structures associated with it.
- Make use of Loader module for video pipeline.
- Remove VideoDecoderConfig, VideoReaderConfig, VideoStorageType and VideoDecoderType
- Modify Sequence Rearrange wrt tensor.
- Fix the brightness and CMN augmentations to operate on sequences

### CODE REVIEW

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 1]
```diff
@@ -1,299 +1,264 @@
...
-    vxReleaseImage(&image);
+
+    // Check for input parameters
+    vx_tensor input;
+    vx_parameter input_param;
+    size_t in_num_tensor_dims;
+    input_param = vxGetParameterByIndex(node, 0);
+    STATUS_ERROR_CHECK(vxQueryParameter(input_param, VX_PARAMETER_ATTRIBUTE_REF, &input, sizeof(vx_tensor)));
+    STATUS_ERROR_CHECK(vxQueryTensor(input, VX_TENSOR_NUMBER_OF_DIMS, &in_num_tensor_dims, sizeof(in_num_tensor_dims)));
+    if(in_num_tensor_dims < 5)
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Does this node support any tensor dims? Or else change the check to the specific case

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 2]
```diff
@@ -1,299 +1,264 @@
...
-    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_IMAGE_FORMAT, &df_image, sizeof(df_image)));
-    vx_uint32 height, width;
-    STATUS_ERROR_CHECK(vxQueryImage(image, VX_IMAGE_ATTRIBUTE_HEIGHT, &height, sizeof(height)));
-    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_IMAGE_HEIGHT, &height, sizeof(height)));
-    STATUS_ERROR_CHECK(vxQueryImage(image, VX_IMAGE_ATTRIBUTE_WIDTH, &width, sizeof(width)));
-    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_IMAGE_WIDTH, &width, sizeof(width)));
-    vxReleaseImage(&image);
+
+    // Check for input parameters
+    vx_tensor input;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: looks like input is not initialized here. Please use the parameter corresponding to input and output

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 3]
```diff
@@ -1,299 +1,264 @@
...
+    vx_parameter input_param;
+    size_t in_num_tensor_dims;
+    input_param = vxGetParameterByIndex(node, 0);
+    STATUS_ERROR_CHECK(vxQueryParameter(input_param, VX_PARAMETER_ATTRIBUTE_REF, &input, sizeof(vx_tensor)));
+    STATUS_ERROR_CHECK(vxQueryTensor(input, VX_TENSOR_NUMBER_OF_DIMS, &in_num_tensor_dims, sizeof(in_num_tensor_dims)));
+    if(in_num_tensor_dims < 5)
+        return ERRMSG(VX_ERROR_INVALID_DIMENSION, "validate: SequenceRearrange: tensor: #0 dimensions=%lu (must be greater than or equal to 4)\n", in_num_tensor_dims);
+
+    // Check for output parameters
+    vx_tensor output;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: output is not initialized

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 4]
```diff
@@ -1,299 +1,264 @@
...
+    STATUS_ERROR_CHECK(vxQueryParameter(output_param, VX_PARAMETER_ATTRIBUTE_REF, &output, sizeof(vx_tensor)));
+    STATUS_ERROR_CHECK(vxQueryTensor(output, VX_TENSOR_NUMBER_OF_DIMS, &out_num_tensor_dims, sizeof(out_num_tensor_dims)));
+    STATUS_ERROR_CHECK(vxQueryTensor(output, VX_TENSOR_DIMS, &tensor_dims, sizeof(tensor_dims)));
+    STATUS_ERROR_CHECK(vxQueryTensor(output, VX_TENSOR_DATA_TYPE, &tensor_type, sizeof(tensor_type)));
+    STATUS_ERROR_CHECK(vxQueryTensor(output, VX_TENSOR_FIXED_POINT_POSITION, &tensor_fixed_point_position, sizeof(tensor_fixed_point_position)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_NUMBER_OF_DIMS, &out_num_tensor_dims, sizeof(out_num_tensor_dims)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_DIMS, &tensor_dims, sizeof(tensor_dims)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_DATA_TYPE, &tensor_type, sizeof(tensor_type)));
+    STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[1], VX_TENSOR_FIXED_POINT_POSITION, &tensor_fixed_point_position, sizeof(tensor_fixed_point_position)));
+    vxReleaseTensor(&input);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: do not release tensor here. Just reference it for querying. See samples in amd_nn extensions for proper usage

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 5]
```diff
@@ -1,299 +1,264 @@
...
+    vxReleaseTensor(&output);
+    vxReleaseParameter(&input_param);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Not required to release parameters and tensors

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp [hunk 6]
```diff
@@ -1,299 +1,264 @@
...
     return VX_SUCCESS;
 }
 
-static vx_status VX_CALLBACK uninitializeSequenceRearrange(vx_node node, const vx_reference *parameters, vx_uint32 num)
-{
+static vx_status VX_CALLBACK uninitializeSequenceRearrange(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    SequenceRearrangeLocalData *data;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    STATUS_ERROR_CHECK(releaseGraphHandle(node, data->handle, data->deviceType));
+    free(data->newOrder);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: please check the data validity before freeing

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 7]
```diff
@@ -1852,28 +1852,22 @@ VX_API_CALL vx_node VX_API_CALL vxExtrppNode_NopbatchPD(vx_graph graph, vx_image
...
-		};
-		 node = createNode(graph, VX_KERNEL_RPP_SEQUENCEREARRANGE, params, 7);
-	}
-	return node;
+VX_API_CALL vx_node VX_API_CALL vxExtrppNode_SequenceRearrange(vx_graph graph, vx_tensor pSrc, vx_tensor pDst, vx_array newOrder, vx_scalar layout) {
+    vx_node node = NULL;
+    vx_context context = vxGetContext((vx_reference)graph);
+    if(vxGetStatus((vx_reference)context) == VX_SUCCESS) {
+        vx_uint32 dev_type = getGraphAffinity(graph);
+        vx_scalar DEV_TYPE = vxCreateScalar(vxGetContext((vx_reference)graph), VX_TYPE_UINT32, &dev_type);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp]: we should stick to same naming convension for all parameter types

File: rocAL/rocAL/include/readers/video/video_reader.h [hunk 8]
```diff
@@ -99,7 +43,7 @@ class VideoReader
     /*!
      \param desc  User provided descriptor containing the files' path.
     */
-    virtual Status initialize(VideoReaderConfig desc) = 0;
+    virtual Status initialize(ReaderConfig desc) = 0;
 
     //! Reads the next resource item
     virtual SequenceInfo get_sequence_info() = 0;
```
Comment (rrawther) [inline rocAL/rocAL/include/readers/video/video_reader.h:53]: I see the SequenceInfo definition is removed and used here

File: rocAL/rocAL/source/api/rocal_api_augmentation.cpp [hunk 9]
```diff
@@ -32,37 +32,32 @@ THE SOFTWARE.
...
+        rocalTensorInfo output_info = input->info();
+        std::vector<size_t> new_dims;
+        new_dims = output_info.dims();
+        new_dims[1] = new_order.size();
+        output_info.set_dims(new_dims);
+
+        output = context->master_graph->create_tensor(output_info, is_output);
         std::shared_ptr<SequenceRearrangeNode> sequence_rearrange_node =  context->master_graph->add_node<SequenceRearrangeNode>({input}, {output});
-        sequence_rearrange_node->init(new_order, new_sequence_length, sequence_length, context->master_graph->internal_batch_size());
+        sequence_rearrange_node->init(new_order);
```
Comment (rrawther) [inline rocAL/rocAL/source/api/rocal_api_augmentation.cpp:58]: why is the sequence_length and new_sequence_length not needed anymore?

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 10]
```diff
@@ -405,12 +403,16 @@ rocalSequenceReaderSingleShard(
...
-        output = context->master_graph->create_loader_output_image(info);
+        std::vector<size_t> dims = {context->user_batch_size(), sequence_length, height, 
+                                    width, static_cast<unsigned>(num_of_planes)};
+        auto info  = rocalTensorInfo(std::move(dims),
+                                     context->master_graph->mem_type(),
+                                     RocalTensorDataType::UINT8);
+        info.set_color_format(color_format);
+        info.set_tensor_layout(RocalTensorlayout::NFHWC);
+        info.set_sequence_batch_size(sequence_length);
+        info.set_max_shape();
```
Comment (rrawther) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:414]: what does set_max_shape() api do here?

File: rocAL/rocAL/source/api/rocal_api_augmentation.cpp (hunk 9 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_augmentation.cpp:58]: Previously we did not have the provision to store sequence length in ImageInfo
Now we can get the sequence length and new sequence length directly from the dims of input tensor and the new order vector length respectively.

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 10 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:414]: Set max shape is to store the max width and height of the tensor wrt the layout passed.
It is set only only for the dataloaders i.e the first node.

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 11]
```diff
@@ -287,16 +287,14 @@ rocalSequenceReader(
         bool shuffle,
         bool loop,
         unsigned step,
-        unsigned stride)
-{
-    Image* output = nullptr;
+        unsigned stride) {
+    rocalTensor* output = nullptr;
```
Comment (rrawther) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:291]: please move this after the context validity check

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 12]
```diff
@@ -342,22 +345,19 @@ rocalSequenceReader(
                                                                             step, stride);
         context->master_graph->set_loop(loop);
 
-        if(is_output)
-        {
+        if(is_output) {
             auto actual_output = context->master_graph->create_image(info, is_output);
```
Comment (rrawther) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: Is this create_image or tensor?

**Review by rrawther — CHANGES_REQUESTED** [review]
@fiona-gladwin : please address review comments

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 11 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:291]: We are checking if the context is set and we throw the error and return the rocalTensor
Hence the output rocalTensor is initialized to nullptr before the validity check

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 1 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 6 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

File: rocAL/rocAL/include/readers/video/video_reader.h (hunk 8 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/readers/video/video_reader.h:53]: Moved back to video_reader.h

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp (hunk 7 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp]: Done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 12 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: Changed to create_tensor

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 2 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 4 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 5 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

File: amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp (hunk 3 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/SequenceRearrange.cpp]: Done

**Comment by fiona-gladwin** [comment]
> @fiona-gladwin : please address review comments

Hi @rrawther I have addressed all review comments. Please check

**Review by rrawther — APPROVED** [review]

**Comment by rrawther** [comment]
@fiona-gladwin : Please merge tot changes before making next PR

**Pull request merged**